Issue 47: client may skip advertising task (#67)
wangshan/draft-wang-ppm-dap-taskprov: draft-wang-ppm-dap-taskprov.md

> OPEN ISSUE: In case of opt-out, would it be useful to specify how to report
> this to the Author?

Once the client opts into a task, it may begin uploading reports for the task
to the Leader. The extension codepoint `taskbind` MUST be offered in the
`extensions` field of both Leader and Helper's `PlaintextInputShare`. In
addition, each report's task ID MUST be computed as described in {{definition}}.

The Client MAY advertise the task configuration by specifying the encoded
`TaskConfig` {{definition}} in the "dap-taskprov" HTTP header, but may choose
to just specify the `taskbind` codepoint in the header, in order to save

> network bandwidth. However, the Leader may respond with "unrecognizedTask" if

Reviewer: Hmmm, this is interesting. Why does client need to send the header at all? What do you expect the leader if it has opted in to the task, but the client doesn't send the header?

Author: The goal is to distinguish between taskprov and non-taskprov tasks. If we don't do this, I guess we can decrypt leader share and try to find the extension codepoint. Both should work. Do you prefer the latter?

Reviewer: If the leader has opted in, then it knows that the task is taskprov. I don't think signaling in the header as well as necessary. Dropping it here would make the protocol simpler, so I'd be in favor of dropping it.

Author: Sounds good. Updated the PR.

> OPEN ISSUE: In case of opt-out, would it be useful to specify how to report
> this to the Author?

Once the client opts into a task, it may begin uploading reports for the task
to the Leader. The extension codepoint `taskbind` MUST be offered in the
`extensions` field of both Leader and Helper's `PlaintextInputShare`. In
addition, each report's task ID MUST be computed as described in {{definition}}.

> The Client MAY advertise the task configuration by specifying the encoded
> `TaskConfig` described in {{definition}} in the "dap-taskprov" HTTP header, but
> may choose to omit this header in order to save network bandwidth. However, the

Reviewer: More perscriptive
```suggestion
The Client SHOULD advertise the task configuration by specifying the encoded
`TaskConfig` described in {{definition}} in the "dap-taskprov" HTTP header, but
MAY choose to omit this header in order to save network bandwidth. However, the
```

Leader may respond with "unrecognizedTask" if it has not been configured with

> this task. In this case, the Client MUST retry the upload request by specifying
> the "dap-taskprov" HTTP header.

Reviewer: ```suggestion
this task. In this case, the Client MUST retry the upload request with
the "dap-taskprov" HTTP header.
```

## Leader Behavior

### Upload Protocol

Upon receiving a Client report, if the Leader does not support the {{taskprov}}
mechanism, it will ignore the "dap-taskprov" HTTP header. In particular, if the

> task ID is notrecognized, then it MUST abort the upload request with

Reviewer: ```suggestion
task ID is not recognized, then it MUST abort the upload request with
```

"unrecognizedTask".

Otherwise, if the Leader does support this mechanism, it first checks if the
"dap-taskprov" HTTP header is specified. If not, that means the Client has
skipped task advertisement. If the Leader recognizes the task ID, it will
include the client report in the aggregation of that task ID. Otherwise, it

> MUST abort with "unrecognizedTask" and ask the Client to retry with the
> "dap-taskprov" HTTP header.

Reviewer: It does not have to "ask" explicitly.
```suggestion
MUST abort with "unrecognizedTask". The Client will then retry with the task advertisement.
```

> If the Client specifies the "dap-taskprov" HTTP header, the Leader checks that

Reviewer: A simpler, and more complete way to say this is
```suggestion
If the Client advertises the task, the Leader checks that
```
